### DISCUSSION TITLE
Add covariance generic to IDataPortal<T> and IChildDataPortal<T>

### DISCUSSION THREAD

**Discussion by hoangcuongvn** [MarimerLLC/csla#2915] [Ideas]
@rockfordlhotka Can you add covariance generic to `IDataPortal<T>` and `IChildDataPortal<T>`. That requires just a little bit change (as the code below) but makes Csla more flexible.
```
public interface IDataPortal<out T>
{
  ...
}

public interface IChildDataPortal<out T>
{
  ...
}
```

E.g in my case, I need to use `IDataPortal<>` through refection as the code below. It only works if `IDataPortal<>` is covariance generic.
```
public object CreateBusinessObject(Type type, params object[] criteria)
{
  var dpType = typeof(IDataPortal<>).MakeGenericType(type);
  
  // This cast requires covariance generic
  var dp = (IDataPortal<object>)_serviceProvider.GetRequiredService(dpType);

  return dp.Create(criteria);
}
```

**Comment by rockfordlhotka** [comment]
From where do you get `_serviceProvider`? I ask, because quite a lot of work goes into ensuring that `AppliationContext` provides you with the correct service provider depending on the runtime hosting the app, and whether the code is running within a local or remote data portal. It is possible that if you are maintaining your own reference to some service provider that you aren't replicating all that behavior and may end up with some very unexpected results.

**Reply by rockfordlhotka** [reply]
> Back to your question. I should use ApplicationContext.GetRequiredService but actually it does not support getting service from a Type. I mean object GetRequiredService(Type type). I hope you could add this variation method too.

Good idea, that would be a helpful override.

> In the meantime, I could use several workarounds

Those workarounds are good. Server-side Blazor is the problem, because it is a unique environment compared to all others thus far.

**Comment by hoangcuongvn** [comment]
@rockfordlhotka My point here is just Csla changes a little bit code to help working with reflection easier. The `_serviceProvider` member is only to illustrate how to get an `IDataPortal<>` instance. The real problem is how to cast an `IDataPortal<>` instance to `IDataPortal<object>` or `IDataPortal<IBusinessObject>`?

Back to your question. I should use `ApplicationContext.GetRequiredService` but actually it does not support getting service from a `Type`. I mean `object GetRequiredService(Type type)`. I hope you could add this variation method too. 

In the meantime, I could use several workarounds. E.g
1. If a Wpf/Console/Worker app, it is easy because it usually does not use scoped services. And we can store an IServiceProvider as a global variable. Or if it uses Caliburn Micro, I will use CM's `IoC` instead.
2. If a MVC/API website, I can get it through IHttpContextAccessor.
3. In the last resort I can add an extension method `public static object GetRequiredService(this ApplicationContext context, Type type)` (I will use relection to call ApplicationContext's non-public methods to implement it).

To be honest, getting a service from a type should not be a problem.

**Reply by swegele** [reply]
Having just gone through the Csal 5 to 6 upgrade on a several large business layers...I did not end up needing what you mention...perhaps because I have made a set of my own base classes that inherit from Csla and which all my business objects inherit from:

**Example:**
```cs
    [Serializable()]
    public abstract class MyChildBusinessListBase<T, C> : MyCommonBusinessListBase<T, C>
        where T : MyChildBusinessListBase<T, C>
        where C : MyChildBusinessBase<C>
    {
        /// <summary>
        /// Gets a dataportal for this list
        /// Use if on client side and data access is needed which requires being on server side (logical or physical)
        /// </summary>
        /// <param name="appContext"></param>
        /// <returns></returns>
        protected static IDataPortal<T> GetDataPortal(ApplicationContext appContext)
        {
            var portalFactory = appContext.GetRequiredService<IDataPortalFactory>();
            var portal = portalFactory.GetPortal<T>();
            return portal;
        }

        /// <summary>
        /// Gets a child data portal for the items in this list
        /// Use if already on server side of dataportal
        /// OR use if no data access needed
        /// </summary>
        /// <param name="appContext"></param>
        /// <returns></returns>
        protected static IChildDataPortal<C> GetItemChildDataPortal(ApplicationContext appContext)
        {
            var portalFactory = appContext.GetRequiredService<IChildDataPortalFactory>();
            var portal = portalFactory.GetPortal<C>();
            return portal;
        }

        /// <summary>
        /// Gets a data portal for the items in this list
        /// Use if on client side and data access is needed which requires being on server side (logical or physical)
        /// </summary>
        /// <param name="appContext"></param>
        /// <returns></returns>
        protected static IDataPortal<C> GetItemDataPortal(ApplicationContext appContext)
        {
            var portalFactory = appContext.GetRequiredService<IDataPortalFactory>();
            var portal = portalFactory.GetPortal<C>();
            return portal;
        }

...
```

Using this technique for each Csla business object type covers 99% of all the scenarios.  I only have a couple where I needed to create a custom dataportal on the fly in the actual implemented class.  So I've not needed the generic reflection technique you are mentioning.

Just sharing different technique...hope it helps.  Ignore if this doesn't apply to you.   Cheers.

**Comment by rockfordlhotka** [comment]
I don't know if this is such a simple change. If I add `out` to the `IDataPortal<T>` interface then all the return values of `Task<T>` become invalid because `T` is covariant.

This is all new to me. I haven't used `out T` before, so don't fully understand the consequences - but it is clearly more than just a simple change to two lines of code.

**Comment by rockfordlhotka** [comment]
I did add the non-generic interface types (#2918), so this is now possible and probably addresses your requirement?

```c#
object CreateBusinessObject(ApplicationContext applicationContext, Type type, params object[] criteria)
{
  var dpType = typeof(IDataPortal<>).MakeGenericType(type);

  // This cast requires covariance generic
  var dp = (IDataPortal)applicationContext.GetRequiredService(dpType);

  return dp.Create(criteria);
}
```

**Reply by hoangcuongvn** [reply]
It was my bad. Adding covariance does not resolve the cast problem.

Non-generic interfaces IDataPortal,  IChildDataPortal would resolve the problem, I do not need that cast 👍
